File: ports-js/semantic-tests.js
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { evalSchemeString } from './scheme.js';

// Get current directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Helper function to check if result matches expected output
function matches(structure, target) {
    if (target && typeof target === 'object' && 'type' in target) {
        return structure instanceof Error;
    }
    return structure === target;
}

// Helper function to safely convert any value to string representation
function valueToString(value) {
    if (typeof value === 'symbol') {
        // Extract the symbol's description
        return Symbol.keyFor(value) ? 
            `Symbol(${Symbol.keyFor(value)})` : 
            value.toString();
    }
    
    if (value === null) return 'null';
    if (value === undefined) return 'undefined';
    
    if (Array.isArray(value)) {
        return `[${value.map(valueToString).join(', ')}]`;
    }
    
    if (value instanceof Error) {
        return `Error: ${value.message}`;
    }
    
    return String(value);
}

// Load test cases
let testTable = [];
try {
    const testsPath1 = join(__dirname, '../ports/lispy-tests.json');
    const testsPath2 = join(__dirname, '../ports/lispy-tests2.json');
    
    const tests1 = JSON.parse(readFileSync(testsPath1, 'utf8'));
    testTable = testTable.concat(tests1);
    
    const tests2 = JSON.parse(readFileSync(testsPath2, 'utf8'));
    testTable = testTable.concat(tests2);
} catch (error) {
    console.error(`Error loading test files: ${error.message}`);
    process.exit(1);
}

// Run tests
for (const entry of testTable) {
    const input = entry.input;
    let evalResult;
    
    try {
        evalResult = evalSchemeString(input);
    } catch (error) {
        evalResult = error;
    }
    
    if (matches(evalResult, entry.expected)) {
        console.log(`✅: ${input}`);
    } else {
        console.log(`❌: ${input} got ${valueToString(evalResult)} instead`);
    }
}

File: ports-rb/semantic-tests.rb
require "json"
require "pathname"

# Helper function to check if result matches expected output
def matches(result, target)
  if target.is_a?(Hash) && target.key?("type")
    return result.is_a?(StandardError)
  end

  if target.is_a?(Array)
    return false unless result.is_a?(Array)
    return false unless result.length == target.length

    match = true
    target.each_with_index do |t, i|
      match &&= matches(result[i], t)
    end
    return match
  end

  if result.is_a?(Symbol)
    return result.to_s == target
  end

  result == target
end

# Get current directory
dirname = File.dirname(__FILE__)

# Load test cases
test_table = []
begin
  tests_path1 = File.join(dirname, "../ports/lispy-tests.json")
  tests_path2 = File.join(dirname, "../ports/lispy-tests2.json")

  tests1 = JSON.parse(File.read(tests_path1))
  test_table.concat(tests1)

  tests2 = JSON.parse(File.read(tests_path2))
  test_table.concat(tests2)
rescue => error
  puts "Error loading test files: #{error.message}"
  exit(1)
end

# Import the scheme module - assuming it's defined in a Ruby file
require_relative "scheme"

# Run tests

total = test_table.length
failed = 0

test_table.each do |entry|
  input = entry["input"]

  begin
    eval_result = Scheme.evaluate_string(input)
  rescue => error
    eval_result = error
  end

  if matches(eval_result, entry["expected"])
    puts "✅: #{input}"
  else
    puts "❌: #{input} got #{eval_result}[#{eval_result.class}] instead of #{entry["expected"]}[#{entry["expected"].class}]"
    failed += 1
  end
end
puts "#{total} tests run, #{failed} failed"
